drivers/hmis/lib/tasks/activity_logs.rb

```diff
@@ -0,0 +1,174 @@
+        return unless payload['action'] == 'show'
+
+        # Track suspicious non-GET requests on blob URLs
+        if payload['method'] != 'GET'
+          @suspicious_rows << [timestamp_raw, message_raw, 'Non-GET method']
+          return
+        end
+
+        # Only process successful S3 redirects (status 302 with location header)
+        return unless payload['status'] == 302 && payload['location'].present?
+
+        path = payload['path'].presence
+        return unless path
+
+        timestamp = parse_timestamp(timestamp_raw)
+        return unless timestamp
+
+        @total_considered += 1
+
+        found = find_matches(timestamp: timestamp, path: path)
+        if found.present?
+          @matched += 1
+        else
+          @unmatched_rows << [timestamp_raw, message_raw]
+        end
+      end
+
+      def parse_payload(message_raw)
+        JSON.parse(message_raw)
+      rescue JSON::ParserError
+        nil
+      end
+
+      def parse_timestamp(timestamp_raw)
+        Time.zone.parse(timestamp_raw)
+      rescue ArgumentError
+        nil
+      end
+
+      def get_client_id_from_path(path)
+        # The path is expected to be in the format:
+        # /rails/active_storage/blobs/redirect/<signed_id>/<filename>
+        # The regex below extracts the signed_id.
+        signed_id = path.match(/\/blobs\/redirect\/([^\/]+)\//)&.captures&.first
+        return unless signed_id.present?
+
+        blob = ActiveStorage::Blob.find_signed(signed_id)
+        return unless blob
+
+        # This logic assumes a blob is only ever attached to one record,
+        # or that the first attachment is always the correct one.
+        attachment = blob.attachments.first
+        return unless attachment&.record.is_a?(Hmis::File)
+
+        attachment.record.client_id
+      rescue ActiveSupport::MessageVerifier::InvalidSignature
+        @invalid_signed_ids += 1
+        nil
+      end
+
+      def client_id_from_log_record(record)
+        # Priority: explicit client header -> header enrollment -> file lookup -> variables
+        return record.header_client_id if record.header_client_id.present?
+
+        ids = extract_ids(record)
+
+        client_id_from_enrollment(record.header_enrollment_id) ||
+          client_id_from_file(ids[:file_id]) ||
+          ids[:client_id] ||
+          client_id_from_enrollment(ids[:enrollment_id])
+      end
+
+      def client_id_from_file(file_id)
+        return unless file_id
+
+        Hmis::File.with_deleted.find_by(id: file_id)&.client_id
+      end
+
+      def client_id_from_enrollment(enrollment_id)
+        return unless enrollment_id
+
+        enrollment = Hmis::Hud::Enrollment.
+          with_deleted.
+          includes(:client_including_deleted, :client).
+          find_by(id: enrollment_id)
+
+        enrollment&.client_including_deleted&.id || enrollment&.client&.id
+      end
+
+      def extract_ids(record)
```

Comment:
This may not all be relevant since we prioritize matching the header enrollment/client ids. This probably could be simplified or removed; maybe we only need GetFile